### PR TITLE
OTT-1886: Drop the EID object if the ID value is empty, to avoid bad requests

### PR DESCRIPTION
**Pull request by Pubmatic-Supriya-Patil**
…requests

# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 1]
```diff
@@ -4495,7 +4498,39 @@ func (o *OpenRTB) ORTBUserExtEIDS() (err error) {
...
+		}
+
+		validUIDs := []map[string]interface{}{}
+		for _, uid := range uids {
+			uidMap, ok := uid.(map[string]interface{})
+			if !ok {
+				continue
+			}
+
+			if id, ok := uidMap["id"].(string); ok && id != "" {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4516]: Please declare ```id``` in constants

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 2]
```diff
@@ -4505,6 +4520,39 @@ func (o *OpenRTB) ORTBUserExtEIDS() (err error) {
 	return
 }
 
+func validateEIDs(eids []map[string]interface{}) []map[string]interface{} {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4523]: NitPick: ```interface{}``` ---> ```any```

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 3]
```diff
@@ -24,6 +25,13 @@ type OpenRTB struct {
 	ortb    *openrtb2.BidRequest
 }
 
+var uidRegexp = regexp.MustCompile(`^(UID2|ID5|BGID|euid|PAIRID|IDL|connectid|firstid|utiq):`)
+
+const (
```
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:30]: Can we move this to constants file

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 4]
```diff
@@ -4495,7 +4503,14 @@ func (o *OpenRTB) ORTBUserExtEIDS() (err error) {
 		return fmt.Errorf(ErrJSONUnmarshalFailed, ORTBUserExtEIDS, "Failed to unmarshal user.ext.eids", eidsValue)
 	}
 
-	userExt[ORTBExtEIDS] = eids
+	validEIDs := validateEIDs(eids)
+
+	if len(validEIDs) == 0 {
+		delete(userExt, ORTBExtEIDS)
+	} else {
+		userExt[ORTBExtEIDS] = validEIDs
```
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4506]: We don't need else here, if no valid eids found we can return it from here

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/constant.go [hunk 5]
```diff
@@ -16,6 +16,11 @@ const (
 	PrebidKey  = "prebid"
 )
 
+const (
+	UIDS = "uids"
```
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/constant.go:20]: Nitpick: Can we follow naming convention same as rest constants in the file

**Review by pm-nikhil-vaidya — CHANGES_REQUESTED** [review]
Added some minor comments

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 6]
```diff
@@ -4505,6 +4515,39 @@ func (o *OpenRTB) ORTBUserExtEIDS() (err error) {
...
 
+func validateEIDs(eids []map[string]interface{}) []map[string]interface{} {
+	validEIDs := []map[string]interface{}{}
+	for _, eid := range eids {
+		uids, ok := eid[UIDS].([]interface{})
+		if !ok {
+			continue
+		}
+
+		validUIDs := []map[string]interface{}{}
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4526]: Use make instead ```validUIDs := make([]map[string]any, len(uids))```

**Review by pm-nikhil-vaidya — APPROVED** [review]

**Review by AvinashKapre — APPROVED** [review]

**Review by pm-viral-vala — APPROVED** [review]

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/util.go [hunk 7]
```diff
@@ -11,3 +13,23 @@ func GetPubIdFromQueryParams(params url.Values) string {
 	}
 	return pubId
 }
+
+func ValidateEIDs(eids []openrtb2.EID) []openrtb2.EID {
+	validEIDs := []openrtb2.EID{}
```
Comment (pm-jaydeep-mohite) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/util.go:18]: can we have 

`validEIDs := make([]openrtb2.EID, 0, len(eids))`

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 4 above)
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4504]: Updated condition

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 8]
```diff
@@ -4495,7 +4498,39 @@ func (o *OpenRTB) ORTBUserExtEIDS() (err error) {
 		return fmt.Errorf(ErrJSONUnmarshalFailed, ORTBUserExtEIDS, "Failed to unmarshal user.ext.eids", eidsValue)
 	}
 
-	userExt[ORTBExtEIDS] = eids
+	// Filter EIDs and UIDs
+	validEIDs := []map[string]interface{}{}
+	for _, eid := range eids {
+		uids, ok := eid["uids"].([]interface{})
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Done

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 1 above)
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Done

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 3 above)
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Done

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/constant.go (hunk 5 above)
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/constant.go]: Done

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 6 above)
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Updated

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/util.go (hunk 7 above)
Comment (Pubmatic-Supriya-Patil) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/util.go:18]: Done

**Review by pm-viral-vala — APPROVED** [review]

**Pull request merged**